src/executionFiles/utils.ts

```diff
@@ -0,0 +1,214 @@
+import {
+  TransactionReceipt,
+  TransactionResponse,
+} from '@ethersproject/providers'
+import { ethers } from 'ethers'
+import { ChainId, ParsedReceipt } from '..'
+import { getArchiveRpcProvider } from '../connectors'
+
+export const defaultReceiptParsing = async (params: {
+  result: ParsedReceipt
+  tx: TransactionResponse
+  receipt: TransactionReceipt
+  toAddress: string
+}) => {
+  let result = params.result
+  const { tx, receipt, toAddress } = params
+
+  // transferred ERC20 in
+  const ercFrom = getTransferredTokenBalance({
+    receipt,
+    fromAddress: toAddress,
+  })
+  if (ercFrom?.toAmount) {
+    result.fromAmount = ercFrom.toAmount
+  }
+
+  // transferred ERC20 out
+  result = {
+    ...result,
+    ...getTransferredTokenBalance({ receipt, toAddress }),
+  }
+
+  // transferred Native Token
+  if (result.toAmount === '0') {
+    result = {
+      ...result,
+      ...(await getTransferredNativeTokenBalance({ tx, receipt, toAddress })),
+    }
+  }
+
+  // gas
+  result = {
+    ...result,
+    ...getUsedGas({ tx, receipt }),
+  }
+  return result
+}
+
+export const getUsedGas = (params: {
+  tx: TransactionResponse
+  receipt: TransactionReceipt
+}) => {
+  const { tx, receipt } = params
+
+  const gasUsed = receipt.gasUsed.toString()
+  const gasPrice = tx.gasPrice?.toString() || '0'
+  const gasFee = receipt.gasUsed.mul(gasPrice).toString()
+
+  return {
+    gasFee,
+    gasPrice,
+    gasUsed,
+  }
+}
+
+export const getTransferredNativeTokenBalance = async (params: {
+  tx: TransactionResponse
+  receipt: TransactionReceipt
+  toAddress: string
+}) => {
+  const { tx, receipt, toAddress } = params
+
+  if (tx.chainId === ChainId.POL && toAddress) {
+    const result = getTransferredNativeTokenBalanceFromLog({
+      receipt,
+      toAddress,
+    })
+    if (result) {
+      return result
+    }
+  }
+
+  return getTransferredNativeTokenBalanceFromChain({ tx, receipt, toAddress })
+}
+
+export const getTransferredNativeTokenBalanceFromChain = async (params: {
+  tx: TransactionResponse
+  receipt: TransactionReceipt
+  toAddress: string
+}) => {
+  const { tx, receipt, toAddress } = params
+
+  // try to load gas balance differences
+  try {
+    const provider = getArchiveRpcProvider(tx.chainId)
+    const providerBlockNumber = await provider.getBlockNumber()
+    if (tx.blockNumber && providerBlockNumber) {
+      const balanceBefore = await provider.getBalance(
+        toAddress,
+        tx.blockNumber - providerBlockNumber - 1
+      )
+      const balanceAfter = await provider.getBalance(
+        toAddress,
+        tx.blockNumber - providerBlockNumber
+      )
+      const balanceDiff = balanceAfter.sub(balanceBefore)
```

Comment:
Is this really necessary? Shouldn't we be able to read the balance from the transaction?
Getting the balance for other than the latest block number was very slow when I tried it

Reply:
I see that this is only called when the `toAmount` is 0. But why does that happen?

Reply:
The `toAmount` is the value we want to find, we usually know what token we expect to get and have an estimate how much, but to get the exact amount we parse the logs.
This function only has to be used when it was not possible to find any other transferred tokens. In this case the user got the native token which does not trigger an event when transferred, instead an internal transaction transfers these tokens.

The only solution I came up to read this from the chain is checking if the gas balance before and after the specific transaction has changed.